### PR TITLE
Introduce Microsoft Fabric SDK

### PR DESCRIPTION
**Pull request by badeamarjieh**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/hdinsight/ci.mgmt.yml
Comment (ArcturusZhang) [inline sdk/hdinsight/ci.mgmt.yml:1]: we should revert this change

File: sdk/resourcemanager/ci.mgmt.yml
Comment (ArcturusZhang) [inline sdk/resourcemanager/ci.mgmt.yml:1]: we should revert this as well

File: sdk/hdinsight/ci.mgmt.yml
Comment (ArcturusZhang) [reply sdk/hdinsight/ci.mgmt.yml:1]: maybe your local branch is not up-to-date with the latest main therefore you are seeing these changes when you run some of the scripts.
Please ensure your branch is up-to-date with the latest main branch

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        public static Azure.ResourceManager.Fabric.Models.RpSkuDetailsForExistingCapacity RpSkuDetailsForExistingCapacity(string resourceType = null, Azure.ResourceManager.Fabric.Models.RpSku sku = null) { throw null; }
+        public static Azure.ResourceManager.Fabric.Models.RpSkuDetailsForNewCapacity RpSkuDetailsForNewCapacity(string resourceType = null, string name = null, System.Collections.Generic.IEnumerable<string> locations = null) { throw null; }
+    }
+    public partial class CheckNameAvailabilityContent : System.ClientModel.Primitives.IJsonModel<Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityContent>, System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityContent>
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:112]: Could we rename this model to `FabricCheckNameAvailabilityContent` to avoid name collision between different libraries?

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityContent System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityContent>.Create(System.BinaryData data, System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+        string System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityContent>.GetFormatFromOptions(System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+        System.BinaryData System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityContent>.Write(System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+    }
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+    public readonly partial struct CheckNameAvailabilityReason : System.IEquatable<Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityReason>
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:124]: same here, please rename this to `FabricCheckNameAvailabilityReason`

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        public override bool Equals(object obj) { throw null; }
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        public override int GetHashCode() { throw null; }
+        public static bool operator ==(Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityReason left, Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityReason right) { throw null; }
+        public static implicit operator Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityReason (string value) { throw null; }
+        public static bool operator !=(Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityReason left, Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityReason right) { throw null; }
+        public override string ToString() { throw null; }
+    }
+    public partial class CheckNameAvailabilityResult : System.ClientModel.Primitives.IJsonModel<Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityResult>, System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.CheckNameAvailabilityResult>
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:141]: same here, please rename this to `FabricCheckNameAvailabilityResult`

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+        void System.ClientModel.Primitives.IJsonModel<Azure.ResourceManager.Fabric.Models.FabricCapacityPatch>.Write(System.Text.Json.Utf8JsonWriter writer, System.ClientModel.Primitives.ModelReaderWriterOptions options) { }
+        Azure.ResourceManager.Fabric.Models.FabricCapacityPatch System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCapacityPatch>.Create(System.BinaryData data, System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+        string System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCapacityPatch>.GetFormatFromOptions(System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+        System.BinaryData System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCapacityPatch>.Write(System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+    }
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+    public readonly partial struct ProvisioningState : System.IEquatable<Azure.ResourceManager.Fabric.Models.ProvisioningState>
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:166]: should rename to `FabricProvisioningState`

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+        public override bool Equals(object obj) { throw null; }
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        public override int GetHashCode() { throw null; }
+        public static bool operator ==(Azure.ResourceManager.Fabric.Models.ProvisioningState left, Azure.ResourceManager.Fabric.Models.ProvisioningState right) { throw null; }
+        public static implicit operator Azure.ResourceManager.Fabric.Models.ProvisioningState (string value) { throw null; }
+        public static bool operator !=(Azure.ResourceManager.Fabric.Models.ProvisioningState left, Azure.ResourceManager.Fabric.Models.ProvisioningState right) { throw null; }
+        public override string ToString() { throw null; }
+    }
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+    public readonly partial struct ResourceState : System.IEquatable<Azure.ResourceManager.Fabric.Models.ResourceState>
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:188]: should rename to `FabricResourceState`

File: sdk/fabric/Azure.ResourceManager.Fabric/src/autorest.md [hunk 6]
```diff
@@ -55,6 +55,11 @@ acronym-mapping:
 
 rename-mapping:
   CheckNameAvailabilityResponse: FabricCheckNameAvailabilityResult
+  CheckNameAvailabilityContent: FabricCheckNameAvailabilityContent
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/src/autorest.md:58]: well some of the renamings are not working.
Please add
```
mgmt-debug:
  show-serialized-names: true
```
and then run `dotnet build /t:GenerateCode` to see the real key to rename them.

**Review by ArcturusZhang — APPROVED** [review]

File: sdk/fabric/test-resources.bicep
Comment (ArcturusZhang) [inline sdk/fabric/test-resources.bicep:1]: as far as I know, this is not how MPG packages handle the test resources.

File: sdk/fabric/Azure.ResourceManager.Fabric/assets.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
+  "AssetsRepo": "Azure/azure-sdk-assets",
+  "AssetsRepoPrefixPath": "net",
+  "TagPrefix": "net/<fabric/Azure.ResourceManager.Fabric",
+  "Tag": ""
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/assets.json:5]: well you did not record your test cases and push the recordings to the recording repo.
Please follow the guidance here to record and push your recordings: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/resourcemanager/Azure.ResourceManager/docs/TestGuide.md

File: sdk/fabric/Azure.ResourceManager.Fabric/tests/Scenario/FabricCapacityTests.cs [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+using System.Net;
+
+namespace Azure.ResourceManager.Fabric.Tests.Scenario
+{
+    public class FabricCapacityTests : FabricManagementTestBase
+    {
+        private FabricCapacityCollection _fabricCapacityCollection { get => ResourceGroup.GetFabricCapacities(); }
+
+        public FabricCapacityTests(bool isAsync)
+            : base(isAsync, RecordedTestMode.Record)
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/tests/Scenario/FabricCapacityTests.cs:27]: this record mode is required when you are recording them, please make sure we remove this record mode when you push your code.
```suggestion
            : base(isAsync)//, RecordedTestMode.Record)
```

File: sdk/fabric/Azure.ResourceManager.Fabric/tests/FabricManagementTestBase.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+            : base(isAsync)
+        {
+        }
+
+        [SetUp]
+        public async Task CreateCommonClient()
+        {
+            Client = GetArmClient();
+            Subscription = await Client.GetDefaultSubscriptionAsync();
+            ResourceGroup = (await Subscription.GetResourceGroupAsync(TestEnvironment.ResourceGroup)).Value;
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/tests/FabricManagementTestBase.cs:40]: is this resource group created by the bicep file?
I am not quite sure if this could work properly during playback.
If it works fine, then it is fine.
If this does not work during playback, we need to create the resource group by our own

File: sdk/fabric/Azure.ResourceManager.Fabric/tests/Scenario/FabricCapacityTests.cs [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+            var capacity = await _fabricCapacityCollection.CreateOrUpdateAsync(WaitUntil.Completed, capacityName, capacityData);
+            return capacity.Value;
+        }
+
+        [TestCase]
+        [RecordedTest]
+        public async Task TestCreateOrUpdate()
+        {
+            // Act
+            var capacityName = "azsdktestcapacity";
```
Comment (ArcturusZhang) [inline sdk/fabric/Azure.ResourceManager.Fabric/tests/Scenario/FabricCapacityTests.cs:46]: We usually suggest we generate the name using
```
Recording.GenerateAssetName("<prefix>")
```

File: sdk/fabric/Azure.ResourceManager.Fabric/tests/FabricManagementTestBase.cs (hunk 9 above)
Comment (ArcturusZhang) [reply sdk/fabric/Azure.ResourceManager.Fabric/tests/FabricManagementTestBase.cs:40]: You could try to playback your recording locally by change the mode to `Playback`.

Comment (badeamarjieh) [reply sdk/fabric/Azure.ResourceManager.Fabric/tests/FabricManagementTestBase.cs:40]: done

**Review by ArthurMa1978 — CHANGES_REQUESTED** [review]
Will migrate to MPG

**Comment by ArcturusZhang** [comment]
Spec update is tracking here: https://github.com/Azure/azure-rest-api-specs/pull/29889

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        public virtual Azure.ResourceManager.ArmOperation<Azure.ResourceManager.Fabric.FabricCapacityResource> Update(Azure.WaitUntil waitUntil, Azure.ResourceManager.Fabric.Models.FabricCapacityPatch patch, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public virtual System.Threading.Tasks.Task<Azure.ResourceManager.ArmOperation<Azure.ResourceManager.Fabric.FabricCapacityResource>> UpdateAsync(Azure.WaitUntil waitUntil, Azure.ResourceManager.Fabric.Models.FabricCapacityPatch patch, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+    }
+    public static partial class FabricExtensions
+    {
+        public static Azure.Response<Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityResult> CheckNameAvailabilityFabricCapacity(this Azure.ResourceManager.Resources.SubscriptionResource subscriptionResource, string location, Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityContent content, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
Comment (ArthurMa1978) [inline sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:64]: Rename to `CheckFabriCapacityNameAvailability`

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+        Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityContent System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityContent>.Create(System.BinaryData data, System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+        string System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityContent>.GetFormatFromOptions(System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+        System.BinaryData System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityContent>.Write(System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+    }
+    public partial class FabricCheckNameAvailabilityResult : System.ClientModel.Primitives.IJsonModel<Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityResult>, System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityResult>
```
Comment (ArthurMa1978) [inline sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:156]: FabricNameAvailabilityResult

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+    {
+        internal FabricCheckNameAvailabilityResult() { }
+        public string Message { get { throw null; } }
+        public bool? NameAvailable { get { throw null; } }
```
Comment (ArthurMa1978) [inline sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:160]: IsNameAvailable

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+        Azure.ResourceManager.Fabric.Models.FabricCapacityProperties System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCapacityProperties>.Create(System.BinaryData data, System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+        string System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCapacityProperties>.GetFormatFromOptions(System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+        System.BinaryData System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCapacityProperties>.Write(System.ClientModel.Primitives.ModelReaderWriterOptions options) { throw null; }
+    }
+    public partial class FabricCheckNameAvailabilityContent : System.ClientModel.Primitives.IJsonModel<Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityContent>, System.ClientModel.Primitives.IPersistableModel<Azure.ResourceManager.Fabric.Models.FabricCheckNameAvailabilityContent>
```
Comment (ArthurMa1978) [inline sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:145]: FabricNameAvailabilityContent

File: sdk/fabric/Azure.ResourceManager.Fabric/src/Generated/Models/RpSkuEnumerationForExistingResourceResult.Serialization.cs [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+using System;
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.Text.Json;
+using Azure.Core;
+
+namespace Azure.ResourceManager.Fabric.Models
+{
+    internal partial class RpSkuEnumerationForExistingResourceResult : IUtf8JsonSerializable, IJsonModel<RpSkuEnumerationForExistingResourceResult>
```
Comment (ArthurMa1978) [inline sdk/fabric/Azure.ResourceManager.Fabric/src/Generated/Models/RpSkuEnumerationForExistingResourceResult.Serialization.cs:16]: uppercase `Rp`

File: sdk/fabric/Azure.ResourceManager.Fabric/src/Generated/Models/CapacityAdministration.cs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Azure.ResourceManager.Fabric.Models
+{
+    /// <summary> The administration properties of the Fabric capacity resource. </summary>
+    public partial class CapacityAdministration
```
Comment (ArthurMa1978) [inline sdk/fabric/Azure.ResourceManager.Fabric/src/Generated/Models/CapacityAdministration.cs:15]: FabricCapacityAdministration

File: sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs (hunk 11 above)
Comment (ArcturusZhang) [reply sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs:64]: The generator current does not support we change this. Created this issue to track: https://github.com/Azure/autorest.csharp/issues/4929

Comment (ArcturusZhang) [reply sdk/fabric/Azure.ResourceManager.Fabric/api/Azure.ResourceManager.Fabric.netstandard2.0.cs]: We decided to postpone this change until we could resolve the above issue.

File: sdk/fabric/Azure.ResourceManager.Fabric/src/Generated/Models/RpSkuEnumerationForExistingResourceResult.Serialization.cs (hunk 15 above)
Comment (ArcturusZhang) [reply sdk/fabric/Azure.ResourceManager.Fabric/src/Generated/Models/RpSkuEnumerationForExistingResourceResult.Serialization.cs:16]: This is a paged result model which will never be public in the future, I think we do not have to change this

**Review by ArcturusZhang — APPROVED** [review]

**Review by ArthurMa1978 — APPROVED** [review]
